projects/graph/src/draw.py

```diff
@@ -1,14 +1,59 @@
 """
 General drawing methods for graphs using Bokeh.
 """
-
-from bokeh.io import show, output_file
-from bokeh.plotting import figure
+import math
+from graph import Graph
+from bokeh.io import show, output_file, show
+from bokeh.plotting import figure, output_file, show
 from bokeh.models import (GraphRenderer, StaticLayoutProvider, Circle, LabelSet,
                           ColumnDataSource)
+from bokeh.palettes import Spectral8
 
-
-class BokehGraph:
+class BokehGraph(Graph):
     """Class that takes a graph and exposes drawing methods."""
-    def __init__(self):
-        pass  # TODO
+    def __init__(self, graph):
+        self.graph = graph
+    def draw(self):
+        graph = self.graph
+        N = 8
+        node_indices = list(range(N))
+
+        plot = figure(title="Graph Layout Demonstration", x_range=(-1.1,1.1), y_range=(-1.1,1.1),
+                    tools="", toolbar_location=None)
+
+        graph = GraphRenderer()
+
+        graph.node_renderer.data_source.add(node_indices, 'index')
+        graph.node_renderer.data_source.add(Spectral8, 'color')
+        graph.node_renderer.glyph = Oval(height=0.1, width=0.2, fill_color="color")
+
+        graph.edge_renderer.data_source.data = dict(
+            start=[0]*N,
+            end=node_indices)
+
+        ### start of layout code
+        circ = [i*2*math.pi/8 for i in node_indices]
+        x = [math.cos(i) for i in circ]
+        y = [math.sin(i) for i in circ]
+        graph_layout = dict(zip(node_indices, zip(x, y)))
+        graph.layout_provider = StaticLayoutProvider(graph_layout=graph_layout)
+
+        ### Draw quadratic bezier paths
+        def bezier(start, end, control, steps):
+            return [(1-s)**2*start + 2*(1-s)*s*control + s**2*end for s in steps]
+
+        xs, ys = [], []
+        sx, sy = graph_layout[0]
+        steps = [i/100. for i in range(100)]
+        for node_index in node_indices:
+            ex, ey = graph_layout[node_index]
+            xs.append(bezier(sx, ex, 0, steps))
+            ys.append(bezier(sy, ey, 0, steps))
+        graph.edge_renderer.data_source.data['xs'] = xs
+        graph.edge_renderer.data_source.data['ys'] = ys
+
+        plot.renderers.append(graph)
+
+        output_file("graph.html")
+        show(plot)
+
```

Comment:
Add this to your draw.py then run the file. then start debugging.
```suggestion
graph = Graph({'0': {'1', '3'},'1': {'0'},'2': set(),'3': {'0'}})

graph.add_vertex(8) #Second Line
graph.add_edge(2, [1,2,3]) #Third line
bg = BokehGraph(graph)
bg.draw()
```